lms/djangoapps/courseware/module_render.py

```diff
@@ -471,29 +484,40 @@ def inner_get_module(descriptor):
             course=course
         )
 
-    def handle_progress_event(block, event_type, event):
+    def get_event_handler(event_type):
         """
-        tie into the CourseCompletions datamodels that are exposed in the edx_solutions_api_integration djangoapp
-        """
-
-        if not settings.FEATURES.get("ALLOW_STUDENT_STATE_UPDATES_ON_CLOSED_COURSE", True):
-            # if a course has ended, don't register progress events
-            course = modulestore().get_course(course_id, depth=0)
-            now = datetime.now(UTC())
-            if course.end is not None and now > course.end:
-                return
+        Return an appropriate function to handle the event.
 
-        user_id = event.get('user_id', user.id)
-        if not user_id:
-            return
+        Returns None if no special processing is required.
+        """
+        handlers = {
+            'completion': handle_completion_event,
+            'grade': handle_grade_event,
+            'progress': handle_deprecated_progress_event,
+        }
+        return handlers.get(event_type)
 
-        CourseModuleCompletion.objects.get_or_create(
-            user_id=user_id,
-            course_id=course_id,
-            content_id=unicode(descriptor.location)
-        )
+    def publish(block, event_type, event):
+        """
+        A function that allows XModules to publish events.
+        """
+        handle_event = get_event_handler(event_type)
+        if handle_event and not is_masquerading_as_specific_student(user, course_id):
+            handle_event(block, event)
+        else:
+            context = contexts.course_context_from_course_id(course_id)
```

Comment:
So I'm not really clear on what happens in this else case here, and why it only happens if there isn't a specific event handler.

But the code matches what was replaced, so 👍

Reply:
Yeah.  It essentially logs the event using the "event_tracker." It's a little hard to trace since "track_function" is dependency-injected here. If I were writing from scratch, I'd probably emit all the events, but as you point out, I was matching the existing behavior.  I don't know if there's an issue with flooding the event tracker with every completed/scored block.

Reply:
Sure, thanks for the explanation!